Newsletters: add 225 (2022-11-09) (#916)
bitcoinops/bitcoinops.github.io: _posts/en/newsletters/2022-11-09-newsletter.md

    responsibly disclosed a vulnerability][topic responsible
    disclosures] that we've summarized in an Optech newsletter.  We
    extend our most heartfelt thanks to everyone listed.  There are
    likely several other disclosures not listed because they haven't
    been made public yet.  We look forward to thanking those discoverers
    and adding them to the topic page of Bitcoin security disclosure
    heroes.  We of course also thank all the reviewers of proposals and
    pull requests whose diligent effort prevented innumerable
    security bugs from making it into released software.

## Bitcoin Core PR Review Club

*In this monthly section, we summarize a recent [Bitcoin Core PR Review Club][]
meeting, highlighting some of the important questions and answers.  Click on a
question below to see a summary of the answer from the meeting.*

[Relax MIN_STANDARD_TX_NONWITNESS_SIZE to 65 non-witness bytes][review club 26265]
is a PR by instagibbs that relaxes the mempool policy's non-witness transaction size
constraints. It allows transactions to be as small as 65 bytes, replacing
the current policy that requires transactions to be at least 85 bytes.

> Since this review club meeting, this PR has been closed in favor of

Reviewer: ```suggestion
Since this Review Club meeting, this PR has been closed in favor of
```

> [PR 26398](https://github.com/bitcoin/bitcoin/pull/26398), which

Reviewer: Perhaps

```suggestion
[Bitcoin Core #26398](https://github.com/bitcoin/bitcoin/pull/26398), which
```

Reviewer: Also, link should be moved to the bottom/reference links

relaxes policy even further by disallowing _only_ 64-byte transactions.
The relative merits of these two slightly-different policies were

> discussed during review club.

Reviewer: ```suggestion
discussed during the meeting.
```

{% include functions/details-list.md
  q0="Why was the minimum transaction size 82 bytes?
  Can you describe the attack?"

>   a0="The 82-byte minimum, which was introduced by PR 11423 in 2018, is

Reviewer: Suggest linking to Bitcoin Core #11423

  the size of the smallest standard payment transaction. This was
  presented as a cleanup of the standardness rules. But in reality,
  the change was to prevent a 64-byte transaction from being considered standard,
  because this size allowed a

>   [spoof payment attack](https://github.com/advisories/GHSA-v55p-4chq-6grj)

Reviewer: Suggest link to https://bitcoinops.org/en/newsletters/2018/12/28/#cve-2017-12842 (and move link to references at the bottom)

  against SPV clients (making them think they've received a payment when they hadn't).
  The attack involves tricking an SPV client into thinking that a 64-byte
  transaction is an inner node of the transaction merkle tree, which is also
  64 bytes in length."

  a0link="https://bitcoincore.reviews/26265#l-35"

>   q1="A participant asked, was it was necessary to fix this vulnerability

Reviewer: Not a huge fan of this question (mostly the answer)

Reviewer 2: I like the question, because it's exactly what I thought before the meeting (when I read how expensive the attack would be). But formulating the answer as another question isn't good (I think that's what you're objecting to); would this be better?
>  a1="There was some agreement, but one participant pointed out that our intuition about this could be wrong."

Or I would be fine with removing the question entirely. It is true that the answer is pretty brief, especially compared to most answers. But really nothing else was said about this question during the meeting.

Author: +1 for keeping this question.  I think it's a really good question given that the vulnerability was kept private for a long time before its accidental disclosure.  We definitely want to encourage developers to look back in hindsight and determine whether that was the best course of action, and be prepared to use their individual conclusions when evaluating future security issues they come across.

Reviewer 2: Good points; this was actually the first time I formulated a question here for the newsletter that wasn't in the host's pre-meeting list of questions. But if a question is asked during the meeting that can provide good content here, seems to me we can include it here.

  a1="There was some agreement, but one participant asked, what if our
  intuition about this is wrong?"

  a1link="https://bitcoincore.reviews/26265#l-66"

  q2="What does _non-witness size_ mean,
  and why do we care about the non-witness distinction?"

  a2="We care about the non-witness distinction because, as part of the
  segwit upgrade, witness data is excluded from the calculation of the merkle root.
  Since the attack requires the malicious transaction to be 64 bytes in the
  merkle root construction (so it looks like an inner node), we need to exclude
  witness data from it."

  a2link="https://bitcoincore.reviews/26265#l-62"

  q3="Why does setting this policy help to prevent the attack?"

  a3="Since inner merkle tree nodes can only be exactly 64 bytes, a transaction
  of a different size cannot be misinterpreted as an inner merkle node."

  a3link="https://bitcoincore.reviews/26265#l-84"

  q4="Does it eliminate the attack vector entirely?"

  a4="Changing the standardness rules only prevents 64-byte transactions
  from being accepted into mempools and relayed, but such transactions
  may still be consensus-valid and so can be mined into a block.

>   For this reason, the attack is still possible, but only by miners."

Reviewer: Perhaps 
```suggestion
  For this reason, the attack is still possible, but only with the help of miners."
```

  a4link="https://bitcoincore.reviews/26265#l-84"

  q5="Why might we want to change the minimum transaction size to 65 bytes,
  apart from the fact that it's unnecessary to obfuscate the CVE?"

  a5="There are legitimate use cases for transactions that are less than 82 bytes.

>   One example mentioned is a CPFP (child pays for parent) transaction that assigns

Reviewer: ```suggestion
  One example mentioned is a [Child Pays For Parent (CPFP)][topic cpfp] transaction that assigns
```

  participate in the public discussion.

## Notable code and documentation changes

*Notable changes this week in [Bitcoin Core][bitcoin core repo], [Core
Lightning][core lightning repo], [Eclair][eclair repo], [LDK][ldk repo],
[LND][lnd repo], [libsecp256k1][libsecp256k1 repo], [Hardware Wallet
Interface (HWI)][hwi repo], [Rust Bitcoin][rust bitcoin repo], [BTCPay
Server][btcpay server repo], [BDK][bdk repo], [Bitcoin Improvement
Proposals (BIPs)][bips repo], and [Lightning BOLTs][bolts repo].*

- [Bitcoin Core #26419][] log: mempool: log removal reason in validation interface FIXME:adamjonas

- [Eclair #2404][] adds support for Short Channel IDentifier (SCID)
  aliases and zero-conf channels (see [Newsletter #156][news156 zc])
  even for channel state commitments that don't use [anchor
  outputs][topic anchor outputs].

- [Eclair #2468][] allows the ultimate receiver of a payment ([HTLC][topic
  HTLC]) to accept a greater amount than they requested and with a
  longer time before it expires than they requested.  Previously,
  Eclair-based receivers adhered to [BOLT4][]'s requirement that the
  amount and expiry delta equal exactly the amount they requested, but

>   that exactitude meant a forwarding node could probe the next hot to

Reviewer: ```suggestion
  that exactitude meant a forwarding node could probe the next hop to
```

  see if it was the final receiver by changing either value by the

>   slightest bit.  [BOLTs #1032][] has been opened to add this privacy

Reviewer: FWIW this has been merged

Author: Thanks!  I'm adding BOLTs 1032 to this newsletter so it's easy for these two PR summaries to reference each other.

       > "-disable_v3_transaction_enforcement" flag to our software, to
       > let users decide to turn off those policy restrictions and
       > treat v3 transactions the same as v2, for all the same reasons
       > that could be argued today with fullrbf [...]
       >
       > [That] would be subversive to making the lightning use case for
       > v3 transactions work [...] we should not enable users to
       > disable this policy, because as long as that policy is just
       > optional and working for those who want it, it shouldn't harm
       > anyone that we offer a tighter set of rules for a particular
       > use case.  Adding a way to bypass those rules is just trying to
       > break someone else's use case, not trying to add a new one.  We
       > should not wield "incentive compatibility" as a bludgeon for
       > breaking things that appear to be working and not causing
       > others harm.
       >
       > I think this is exactly what is happening with fullrbf.

    Daftuar ends his email with three questions for those who still want
    the `mempoolfullrbf` option to be included in Bitcoin Core:

>     1. "Does fullrbf offer any benefits other than breaking zeroconf

Reviewer 2: Should this be zero-conf? We use that below. I made PR #922 to add this to the style guide.

Author: I'm going to leave this as-is since it's in a quote, it's clear, and I don't think it distracts too much that he uses one spelling and we use a different spelling elsewhere in this newsletter.

Reviewer 2: Ah yes, I didn't notice it was in a quote. I agree.
